File: backend/app/db/models/career_coach.py
from __future__ import annotations

import uuid
from datetime import datetime

from sqlalchemy import Boolean, DateTime, ForeignKey, Integer, String, Text, func
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base

PROFICIENCY_LEVELS = ("beginner", "intermediate", "advanced", "expert")
SKILL_SOURCES = ("manual", "resume", "ats", "interview")
ROADMAP_STATUSES = ("draft", "active", "completed", "archived")
MILESTONE_STATUSES = ("pending", "in_progress", "completed", "skipped")


class UserGoal(Base):
    __tablename__ = "user_goals"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    user_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    target_role: Mapped[str] = mapped_column(String(128), nullable=False)
    target_timeline_months: Mapped[int | None] = mapped_column(Integer, nullable=True)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    is_active: Mapped[bool] = mapped_column(Boolean, nullable=False, server_default="true")
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False,
    )

    roadmaps: Mapped[list["CareerRoadmap"]] = relationship(
        "CareerRoadmap",
        back_populates="goal",
        lazy="selectin",
        cascade="all, delete-orphan",
    )


class UserSkill(Base):
    __tablename__ = "user_skills"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    user_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    skill_name: Mapped[str] = mapped_column(String(128), nullable=False)
    proficiency_level: Mapped[str] = mapped_column(String(32), nullable=False, server_default="beginner")
    source: Mapped[str] = mapped_column(String(32), nullable=False, server_default="manual")
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False,
    )


class CareerRoadmap(Base):
    __tablename__ = "career_roadmaps"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    user_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    goal_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("user_goals.id", ondelete="SET NULL"),
        nullable=True,
        index=True,
    )
    title: Mapped[str] = mapped_column(String(256), nullable=False)
    target_role: Mapped[str] = mapped_column(String(128), nullable=False)
    milestones: Mapped[list] = mapped_column(JSONB, nullable=False, server_default="[]")
    status: Mapped[str] = mapped_column(String(32), nullable=False, server_default="active")
    progress_percent: Mapped[int] = mapped_column(Integer, nullable=False, server_default="0")
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False,
    )

    goal: Mapped["UserGoal | None"] = relationship("UserGoal", back_populates="roadmaps")
    progress_entries: Mapped[list["ProgressTracking"]] = relationship(
        "ProgressTracking",
        back_populates="roadmap",
        lazy="selectin",
        cascade="all, delete-orphan",
    )


class ProgressTracking(Base):
    __tablename__ = "progress_tracking"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    user_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    roadmap_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("career_roadmaps.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    milestone_id: Mapped[str] = mapped_column(String(64), nullable=False)
    status: Mapped[str] = mapped_column(String(32), nullable=False, server_default="pending")
    notes: Mapped[str | None] = mapped_column(Text, nullable=True)
    completed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False,
    )

    roadmap: Mapped["CareerRoadmap"] = relationship("CareerRoadmap", back_populates="progress_entries")


class ReadinessScore(Base):
    __tablename__ = "readiness_scores"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    user_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    overall_score: Mapped[int] = mapped_column(Integer, nullable=False)
    category_scores: Mapped[dict] = mapped_column(JSONB, nullable=False, server_default="{}")
    weak_areas: Mapped[list] = mapped_column(JSONB, nullable=False, server_default="[]")
    missing_skills: Mapped[list] = mapped_column(JSONB, nullable=False, server_default="[]")
    recommendations: Mapped[list] = mapped_column(JSONB, nullable=False, server_default="[]")
    computed_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now(), nullable=False)

File: frontend/src/types/career-coach.ts
export interface UserGoal {
  id: string;
  target_role: string;
  target_timeline_months: number | null;
  description: string | null;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface UserSkill {
  id: string;
  skill_name: string;
  proficiency_level: string;
  source: string;
  created_at: string;
  updated_at: string;
}

export interface RoadmapMilestone {
  id: string;
  title: string;
  description: string;
  order: number;
  estimated_weeks: number;
  skills: string[];
  status: string;
}

export interface CareerRoadmap {
  id: string;
  goal_id: string | null;
  title: string;
  target_role: string;
  milestones: RoadmapMilestone[];
  status: string;
  progress_percent: number;
  created_at: string;
  updated_at: string;
}

export interface SkillGapItem {
  skill_name: string;
  priority: string;
  reason: string;
}

export interface SkillGapAnalysis {
  target_role: string;
  required_skills: string[];
  user_skills: string[];
  missing_skills: SkillGapItem[];
  coverage_percent: number;
}

export interface LearningRecommendation {
  title: string;
  description: string;
  skill: string;
  resource_type: string;
  priority: string;
}

export interface WeaknessItem {
  area: string;
  severity: string;
  description: string;
  suggested_action: string;
}

export interface ReadinessScore {
  id: string;
  overall_score: number;
  category_scores: Record<string, number>;
  weak_areas: WeaknessItem[];
  missing_skills: string[];
  recommendations: LearningRecommendation[];
  computed_at: string;
}

export interface CareerCoachDashboard {
  readiness_score: number | null;
  readiness_trend: string | null;
  missing_skills: string[];
  roadmap_progress_percent: number | null;
  weak_areas: WeaknessItem[];
  recommendations: LearningRecommendation[];
  active_goal: UserGoal | null;
  skill_coverage_percent: number | null;
}

export interface UserGoalCreatePayload {
  target_role: string;
  target_timeline_months?: number | null;
  description?: string | null;
}

export interface ProgressUpdatePayload {
  status: string;
  notes?: string | null;
}
